File: api/app/services/m3u_parser.py
"""Interface M3UParser sobre ipytv (m3u-ipytv) — troca de biblioteca não
deve afetar o resto do importador (research.md R1).

Também detecta manifesto HLS (RFC 8216) antes de classificar, para nunca
importar segmentos de streaming como se fossem canais de catálogo (FR-009;
ADR-005 §2).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from ipytv import playlist as ipytv_playlist

# Tags que só existem em manifestos HLS de segmentos/variantes de verdade
# (RFC 8216): TARGETDURATION/MEDIA-SEQUENCE/ENDLIST são obrigatórias/típicas
# de uma media playlist; STREAM-INF/I-FRAME-STREAM-INF marcam uma master
# playlist. Deliberadamente NÃO inclui #EXT-X-SESSION-DATA sozinha — vários
# painéis Xtream/XUI injetam essa tag de branding/versão (ex.:
# "com.xui.1_5_5r2") em catálogos M3U comuns, sem o arquivo ser um manifesto
# de streaming; tratar qualquer "#EXT-X-*" como HLS gerava falso positivo
# rejeitando catálogos reais inteiros.
_HLS_TAG_PATTERN = re.compile(
    r"^#EXT-X-(STREAM-INF|I-FRAME-STREAM-INF|TARGETDURATION|MEDIA-SEQUENCE|"
    r"ENDLIST|DISCONTINUITY|KEY|MAP|BYTERANGE|PLAYLIST-TYPE)\b",
    re.MULTILINE,
)


class HLSManifestDetectedError(ValueError):
    """O conteúdo é um manifesto de streaming (HLS), não uma playlist de catálogo."""


class EmptyPlaylistError(ValueError):
    """A lista M3U é sintaticamente válida mas não tem nenhuma entrada."""


class InvalidPlaylistError(ValueError):
    """O conteúdo não é reconhecível como M3U (ex.: HTML de erro, arquivo truncado)."""


@dataclass(frozen=True)
class ParsedEntry:
    name: str
    url: str
    group: str | None
    attributes: dict[str, str] = field(default_factory=dict)


@dataclass(frozen=True)
class ParseResult:
    entries: list[ParsedEntry]
    invalid_count: int


def is_hls_manifest(text: str) -> bool:
    """HLS reaproveita a tag #EXTM3U, mas só as tags de segmento/variante
    (TARGETDURATION, STREAM-INF, etc. — RFC 8216 seção 4) indicam de fato um
    manifesto de streaming, não um catálogo IPTV comum."""
    return bool(_HLS_TAG_PATTERN.search(text))


class M3UParser:
    def parse(self, text: str) -> ParseResult:
        stripped = text.strip()
        if not stripped:
            raise EmptyPlaylistError("Conteúdo vazio.")
        if not stripped.startswith("#EXTM3U"):
            raise InvalidPlaylistError(
                "Conteúdo não começa com #EXTM3U — não parece uma lista M3U."
            )
        if is_hls_manifest(text):
            raise HLSManifestDetectedError(
                "Conteúdo é um manifesto de streaming (HLS), não uma playlist de catálogo."
            )

        try:
            parsed = ipytv_playlist.loads(text)
        except Exception as exc:  # ipytv levanta exceções próprias (MalformedPlaylistException etc.)
            raise InvalidPlaylistError(f"Falha ao interpretar a lista M3U: {exc}") from exc

        entries: list[ParsedEntry] = []
        invalid_count = 0
        for channel in parsed.get_channels():
            if not channel.url:
                invalid_count += 1
                continue
            group = channel.attributes.get("group-title") or None
            entries.append(
                ParsedEntry(
                    name=channel.name or channel.url,
                    url=channel.url,
                    group=group,
                    attributes=dict(channel.attributes),
                )
            )

        if not entries:
            raise EmptyPlaylistError("Lista M3U sem nenhuma entrada válida.")
        return ParseResult(entries=entries, invalid_count=invalid_count)

File: tv-web/src/lib/catalog/m3uParser.ts
/**
 * Porta de `api/app/services/m3u_parser.py` para o aparelho.
 *
 * Diferença deliberada em relação ao original: aqui o conteúdo é
 * consumido **em fluxo** (D-002). O backend podia carregar o texto
 * inteiro na memória de um servidor; numa TV isso é a diferença entre
 * funcionar e o app fechar sozinho com uma lista de centenas de milhares
 * de entradas.
 *
 * O parser é uma função pura sobre linhas — não conhece rede nem
 * armazenamento. É isso que permite testá-lo sem Worker e sem IndexedDB,
 * e que mantém o invólucro de Worker trocável (D-003).
 */

export interface ParsedEntry {
  name: string
  url: string
  group?: string
  attributes: Record<string, string>
}

export class HlsManifestDetectedError extends Error {
  constructor() {
    super('Conteúdo é um manifesto de streaming (HLS), não uma playlist de catálogo.')
    this.name = 'HlsManifestDetectedError'
  }
}

export class InvalidPlaylistError extends Error {
  constructor(message = 'Conteúdo não parece uma lista M3U.') {
    super(message)
    this.name = 'InvalidPlaylistError'
  }
}

export class EmptyPlaylistError extends Error {
  constructor(message = 'Lista M3U sem nenhuma entrada válida.') {
    super(message)
    this.name = 'EmptyPlaylistError'
  }
}

/**
 * Tags que só existem em manifestos HLS de verdade (RFC 8216).
 *
 * Deliberadamente NÃO inclui `#EXT-X-SESSION-DATA`: vários painéis
 * Xtream/XUI injetam essa tag de branding em catálogos comuns, e tratá-la
 * como HLS reprovava catálogos reais inteiros. Esse falso positivo já foi
 * corrigido uma vez no backend — a regra é portada com a correção junto,
 * não sem ela.
 */
const HLS_TAG = /^#EXT-X-(STREAM-INF|I-FRAME-STREAM-INF|TARGETDURATION|MEDIA-SEQUENCE|ENDLIST|DISCONTINUITY|KEY|MAP|BYTERANGE|PLAYLIST-TYPE)\b/

const ATTRIBUTE = /([\w-]+)="([^"]*)"/g

export function isHlsTagLine(line: string): boolean {
  return HLS_TAG.test(line.trim())
}

/**
 * Separa a parte de duração/atributos do nome do canal.
 *
 * O nome vem depois da primeira vírgula **fora de aspas** — um atributo
 * pode conter vírgula dentro do valor (`group-title="Ação, Aventura"`),
 * e cortar na primeira vírgula bruta quebraria esses casos.
 */
function splitExtinf(payload: string): { head: string; name: string } {
  let inQuotes = false
  for (let i = 0; i < payload.length; i += 1) {
    const char = payload[i]
    if (char === '"') inQuotes = !inQuotes
    else if (char === ',' && !inQuotes) {
      return { head: payload.slice(0, i), name: payload.slice(i + 1).trim() }
    }
  }
  return { head: payload, name: '' }
}

function parseAttributes(head: string): Record<string, string> {
  const attributes: Record<string, string> = {}
  ATTRIBUTE.lastIndex = 0
  let match = ATTRIBUTE.exec(head)
  while (match !== null) {
    attributes[match[1]] = match[2]
    match = ATTRIBUTE.exec(head)
  }
  return attributes
}

export interface ParseTally {
  /** Entradas com `#EXTINF` que não puderam virar item (ex.: sem URL). */
  invalidCount: number
}

/**
 * Interpreta linhas de M3U em fluxo, emitindo uma entrada por vez.
 *
 * Lança `InvalidPlaylistError` se a primeira linha útil não for `#EXTM3U`,
 * e `HlsManifestDetectedError` assim que encontrar uma tag de manifesto.
 *
 * Nota sobre a detecção de HLS: o backend olhava o texto inteiro antes de
 * interpretar qualquer coisa. Em fluxo isso é impossível sem desfazer o
 * ganho de memória, então a tag é detectada **durante** a passagem. O
 * efeito para quem usa é o mesmo: a importação aborta e nada é publicado,
 * porque o pipeline só publica uma geração completa (D-004).
 */
export async function* parseM3uLines(
  lines: AsyncIterable<string>,
  tally: ParseTally = { invalidCount: 0 },
): AsyncGenerator<ParsedEntry> {
  let sawHeader = false
  let pending: { name: string; group?: string; attributes: Record<string, string> } | null = null

  for await (const rawLine of lines) {
    // Remove BOM da primeira linha e espaços das demais.
    const line = rawLine.replace(/^﻿/, '').trim()
    if (line === '') continue

    if (!sawHeader) {
      if (!line.startsWith('#EXTM3U')) {
        throw new InvalidPlaylistError('Conteúdo não começa com #EXTM3U.')
      }
      sawHeader = true
      continue
    }

    if (isHlsTagLine(line)) throw new HlsManifestDetectedError()

    if (line.startsWith('#EXTINF:')) {
      // Um #EXTINF seguido de outro, sem URL no meio, é entrada inválida.
      if (pending !== null) tally.invalidCount += 1
      const { head, name } = splitExtinf(line.slice('#EXTINF:'.length))
      const attributes = parseAttributes(head)
      pending = { name, group: attributes['group-title'] || undefined, attributes }
      continue
    }

    // Outras linhas de diretiva não interessam ao catálogo.
    if (line.startsWith('#')) continue

    if (pending === null) continue // URL solta, sem #EXTINF antes.

    yield {
      name: pending.name || line,
      url: line,
      group: pending.group,
      attributes: pending.attributes,
    }
    pending = null
  }

  // Um #EXTINF no fim do arquivo, sem URL, também é entrada inválida.
  if (pending !== null) tally.invalidCount += 1
  if (!sawHeader) throw new InvalidPlaylistError('Conteúdo vazio.')
}

/** Transforma um texto completo num fluxo de linhas — usado em teste e em listas pequenas. */
export async function* linesFromText(text: string): AsyncGenerator<string> {
  for (const line of text.split(/\r?\n/)) yield line
}

/**
 * Linhas de um corpo de resposta HTTP, sem nunca ter o arquivo inteiro na
 * memória.
 *
 * É o que separa este caminho de um `await response.text()`: uma lista de
 * 300 mil entradas passa dos 50 MB de texto, e materializá-la antes de
 * interpretar já custaria, sozinha, mais memória do que a TV tem para dar
 * (D-002). Aqui o pico é um pedaço de rede mais uma linha.
 *
 * O decodificador é `stream: true` de propósito: um caractere multibyte
 * pode ficar partido entre dois pedaços, e decodificar cada pedaço isolado
 * corromperia acentos exatamente nos nomes de canal em português.
 */
export async function* linesFromResponse(body: ReadableStream<Uint8Array>): AsyncGenerator<string> {
  const reader = body.getReader()
  const decoder = new TextDecoder('utf-8')
  let buffer = ''
  try {
    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })
      let newline = buffer.indexOf('\n')
      while (newline !== -1) {
        yield buffer.slice(0, newline).replace(/\r$/, '')
        buffer = buffer.slice(newline + 1)
        newline = buffer.indexOf('\n')
      }
    }
    buffer += decoder.decode()
    if (buffer !== '') yield buffer.replace(/\r$/, '')
  } finally {
    // Abandonar o fluxo no meio (cancelamento, erro de escrita) tem que
    // soltar a conexão, ou ela fica pendurada até o app fechar.
    reader.releaseLock()
  }
}

export interface ParseResult {
  entries: ParsedEntry[]
  invalidCount: number
}

/**
 * Conveniência para texto completo. **Materializa tudo em memória** — não
 * usar no caminho de importação de lista grande, que deve consumir
 * `parseM3uLines` diretamente (D-002).
 */
export async function parseM3uText(text: string): Promise<ParseResult> {
  const tally: ParseTally = { invalidCount: 0 }
  const entries: ParsedEntry[] = []
  for await (const entry of parseM3uLines(linesFromText(text), tally)) entries.push(entry)
  if (entries.length === 0) throw new EmptyPlaylistError()
  return { entries, invalidCount: tally.invalidCount }
}
